plugins/common/starlark/starlark.go

```diff
@@ -0,0 +1,154 @@
+package starlark //nolint - Needed to avoid getting import-shadowing: The name 'starlark' shadows an import name (revive)
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"go.starlark.net/lib/math"
+	"go.starlark.net/lib/time"
+	"go.starlark.net/resolve"
+	"go.starlark.net/starlark"
+	"go.starlark.net/starlarkjson"
+)
+
+type StarlarkCommon struct {
+	Source    string                 `toml:"source"`
+	Script    string                 `toml:"script"`
+	Constants map[string]interface{} `toml:"constants"`
+
+	Log              telegraf.Logger `toml:"-"`
+	StarlarkLoadFunc func(module string, logger telegraf.Logger) (starlark.StringDict, error)
+
+	thread *starlark.Thread
+}
+
+func (s *StarlarkCommon) Init() (starlark.StringDict, error) {
+	if s.Source == "" && s.Script == "" {
+		return nil, errors.New("one of source or script must be set")
+	}
+	if s.Source != "" && s.Script != "" {
+		return nil, errors.New("both source or script cannot be set")
+	}
+
+	s.thread = &starlark.Thread{
+		Print: func(_ *starlark.Thread, msg string) { s.Log.Debug(msg) },
+		Load: func(thread *starlark.Thread, module string) (starlark.StringDict, error) {
+			return s.StarlarkLoadFunc(module, s.Log)
+		},
+	}
+
+	builtins := starlark.StringDict{}
+	builtins["Metric"] = starlark.NewBuiltin("Metric", newMetric)
+	builtins["deepcopy"] = starlark.NewBuiltin("deepcopy", deepcopy)
+	builtins["catch"] = starlark.NewBuiltin("catch", catch)
+	s.addConstants(&builtins)
+
+	program, err := s.sourceProgram(builtins, "")
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Execute source
+	globals, err := program.Init(s.thread, builtins)
+	if err != nil {
+		return nil, err
+	}
+	// Make available a shared state to the apply function
+	globals["state"] = starlark.NewDict(0)
+
+	// Freeze the global state.  This prevents modifications to the processor
+	// state and prevents scripts from containing errors storing tracking
+	// metrics.  Tasks that require global state will not be possible due to
+	// this, so maybe we should relax this in the future.
+	globals.Freeze()
+
+	return globals, nil
+}
+
+func InitFunction(globals starlark.StringDict, fnName string, expectedParams int) (*starlark.Function, error) {
+	globalFn := globals[fnName]
+
+	if globalFn == nil {
+		return nil, fmt.Errorf("%s is not defined", fnName)
+	}
+
+	var ok bool
+	var fn *starlark.Function
+	if fn, ok = globalFn.(*starlark.Function); !ok {
+		return nil, fmt.Errorf("%s is not a function", fnName)
+	}
+
+	if fn.NumParams() != expectedParams {
+		return nil, fmt.Errorf("%s function must take %d parameter(s)", fnName, expectedParams)
+	}
+	return fn, nil
+}
```

Comment:
How about calling this `AddFunction` and extend the scope. You could do
```suggestion
func (s *StarlarkCommon) AddFunction(name string, params ...starlark.Tuple) (*starlark.Function, starlark.Tuple, error) {
	if _, found := s.globals[name]; !found {
		return nil, nil, fmt.Errorf("%s is not defined", name)
	}

	fn, ok := globalFn.(*starlark.Function)
	if !ok {
		return nil, nil, fmt.Errorf("%s is not a function", fnName)
	}

	if fn.NumParams() != len(params) {
		return nil, nil, fmt.Errorf("%s function must take %d parameter(s)", name, len(params))
	}
	
	p := make(starlark.Tuple, len(params))
	for i, param := range params {
		p[i] = param
	}
	return fn, p, nil
}
```

Reply:
Or even better, make a `functions` and a `parameters` map in the struct and do
How about calling this `AddFunction` and extend the scope. You could do
```suggestion
func (s *StarlarkCommon) AddFunction(name string, params ...starlark.Tuple) error {
	if _, found := s.globals[name]; !found {
		return fmt.Errorf("%s is not defined", name)
	}

	fn, ok := globalFn.(*starlark.Function)
	if !ok {
		return fmt.Errorf("%s is not a function", name)
	}

	if fn.NumParams() != len(params) {
		return nil, nil, fmt.Errorf("%s function must take %d parameter(s)", name, len(params))
	}
	
	s.functions[name] = fn
	s.parameters[name] = params

	return nil
}
```

Reply:
Hum, I'm not really a big fan of this map approach, using a map data structure just to store one element for the processor and three for the aggregator sounds like overkill, don't you agree?

Reply:
Yeah it's kind of overkill, but it moves more code into common, making it easier for people to reuse it. There is e.g. a discussion on scripting http requests/responses which could also use Starlark and maybe would need to add a bunch of functions.
And last but not least, look at the beauty of this `AddFunction` above! ;-P

Reply:
ok then, I'll add the maps